pkg/storage/tsdb/bucketindex/reader.go

```diff
@@ -0,0 +1,55 @@
+package bucketindex
+
+import (
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"io/ioutil"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+)
+
+var (
+	ErrIndexNotFound  = errors.New("bucket index not found")
+	ErrIndexCorrupted = errors.New("bucket index corrupted")
+)
+
+// ReadIndex reads, parses and returns a bucket index from the bucket.
+func ReadIndex(ctx context.Context, bkt objstore.Bucket, userID string, logger log.Logger) (*Index, error) {
+	bkt = bucket.NewUserBucketClient(userID, bkt)
+
+	// Get the bucket index.
+	reader, err := bkt.Get(ctx, IndexCompressedFilename)
+	if err != nil {
+		if bkt.IsObjNotFoundErr(err) {
+			return nil, ErrIndexNotFound
+		}
+		return nil, errors.Wrap(err, "read bucket index")
+	}
+	defer runutil.CloseWithLogOnErr(logger, reader, "close bucket index reader")
+
+	// Read all the content.
+	gzipReader, err := gzip.NewReader(reader)
+	if err != nil {
+		return nil, ErrIndexCorrupted
+	}
+	defer runutil.CloseWithLogOnErr(logger, gzipReader, "close bucket index gzip reader")
```

Comment:
Comment: No need to close gzip reader, since we're ignoring its error anyway. It only checks if final checksum was correct. We would already detect error (eg. json parsing) earlier.

Reply:
I think it's more future proof keeping the call to `.Close()`.